### PR TITLE
Translate legacy tests for buffer operator

### PR DESCRIPTION
**Pull request by shariarriday**
Updates the legacy tests for merge operator to use the new test framework and deterministic framework

### CODE REVIEW

File: libcaf_core/caf/flow/op/buffer.test.cpp [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+    auto ctrl_sub = make_counted<passive_subscription_impl>(coordinator());
+    uut->fwd_on_subscribe(fwd_ctrl,
+                          caf::flow::subscription{std::move(ctrl_sub)});
+  }
+
+  template <class T>
+  auto trivial_obs() {
+    return flow::make_trivial_observable<T>(coordinator());
+  }
+};
```
Comment (Neverlord) [inline libcaf_core/caf/flow/op/buffer.test.cpp:97]: Why do we need two separate fixtures? Did you try something like this instead?

```c++
struct fixture : test::fixture::deterministic, test::fixture::flow {};
```

File: libcaf_core/caf/flow/op/buffer.test.cpp [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+          cow_vector<int>{64, 128},
+        };
+        ctx->make_observable()
+          .from_container(inputs) //
+          .buffer(3)
+          .for_each([&outputs](const cow_vector<int>& xs) {
+            outputs.emplace_back(xs);
+          });
+        ctx->run();
+        check_eq(outputs, expected);
```
Comment (Neverlord) [inline libcaf_core/caf/flow/op/buffer.test.cpp:119]: We have `collect()` now. Why didn't you use it here?

File: libcaf_core/caf/flow/op/buffer.test.cpp [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+          pub.as_observable()
+            .observe_on(self) //
+            .buffer(3, 1s)
+            .do_on_complete([closed] { *closed = true; })
+            .for_each([outputs](const cow_vector<int>& xs) {
+              outputs->emplace_back(xs);
+            });
+        });
+        dispatch_messages();
+        print_info("emit the first six items");
```
Comment (Neverlord) [inline libcaf_core/caf/flow/op/buffer.test.cpp:147]: I think `debug` level would be more appropriate.

**Review by Neverlord — CHANGES_REQUESTED** [review]
Just a first pass, I'll go through the remaining tests after the first round of changes.

File: libcaf_core/caf/flow/op/buffer.test.cpp [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+      = make_counted<test::fixture::flow::passive_subscription_impl>(
+        coordinator());
+    uut->fwd_on_subscribe(fwd_ctrl,
+                          caf::flow::subscription{std::move(ctrl_sub)});
+  }
+
+  template <class T>
+  auto trivial_obs() {
+    return test::fixture::flow::make_trivial_observable<T>(coordinator());
+  }
```
Comment (Neverlord) [inline libcaf_core/caf/flow/op/buffer.test.cpp:63]: I wonder if we really need this. I think the tests would work just as well with `never` and wouldn't require us to hand-roll a new observable implementation for testing.

File: libcaf_core/caf/flow/op/buffer.test.cpp [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+      = make_counted<test::fixture::flow::passive_subscription_impl>(
+        coordinator());
+    uut->fwd_on_subscribe(fwd_data,
+                          caf::flow::subscription{std::move(data_sub)});
+    auto ctrl_sub
+      = make_counted<test::fixture::flow::passive_subscription_impl>(
+        coordinator());
+    uut->fwd_on_subscribe(fwd_ctrl,
+                          caf::flow::subscription{std::move(ctrl_sub)});
+  }
```
Comment (Neverlord) [inline libcaf_core/caf/flow/op/buffer.test.cpp:58]: Did you check whether this still actually does anything after getting rid of `nil_observable`? I believe we can just rip this out as well.

File: libcaf_core/caf/flow/op/buffer.test.cpp [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+      THEN("the observer receives an on_complete event") {
+        auto snk = flow::make_passive_observer<cow_vector<int>>();
+        auto uut = raw_sub<skip_trait>(3, trivial_obs<int>(),
+                                       trivial_obs<int64_t>(),
+                                       snk->as_observer());
+        add_subs(uut);
+        snk->request(42);
+        run_flows();
+        uut->dispose();
+        run_flows();
```
Comment (Neverlord) [inline libcaf_core/caf/flow/op/buffer.test.cpp:460]: I think this is the better (and safer) option to clean up tests. Please drop the unsubscribe guard and use this instead. Can be a convenience function in the fixture or just add these two lines after each test setup.

With this gone as well, I believe the fixture should remain unchanged. Which is a good thing, we had quite a bit of unnecessary complexity in there that we should actively strive to get rid of.

**Review by Neverlord — CHANGES_REQUESTED** [review]

File: libcaf_core/caf/flow/op/buffer.test.cpp (hunk 6 above)
Comment (shariarriday) [reply libcaf_core/caf/flow/op/buffer.test.cpp:447]: We still have `passive_subscription_impl` that was added to this fixture. Should I look into removing it from the fixture as well?

Comment (Neverlord) [reply libcaf_core/caf/flow/op/buffer.test.cpp:447]: Yep, give it a shot. Less is more.

**Review by Neverlord — APPROVED** [review]

**Pull request merged**